File: 1799-maximize-score-after-n-operations/1799-maximize-score-after-n-operations.cpp
class Solution {
public:
    int dp[8][16384] = {};
    int maxScore(vector<int>& n, int i = 1, int mask = 0) {
        if (i > n.size() / 2)
            return 0;
        if (!dp[i][mask])
            for (int j = 0; j < n.size(); ++j)
                for (auto k = j + 1; k < n.size(); ++k) {
                    int new_mask = (1 << j) + (1 << k);
                    if ((mask & new_mask) == 0)
                        dp[i][mask] = max(dp[i][mask], i * __gcd(n[j], n[k]) + maxScore(n, i + 1, mask + new_mask));
                }
        return dp[i][mask];
    }
};

File: 1799-maximize-score-after-n-operations/1799-maximize-score-after-n-operations.py
class Solution:
    def maxScore(self, nums: List[int]) -> int:
        max_score = 0
        
        def dfs(remaining, temp):
            if len(remaining) == 2:
                temp = sorted(temp + [gcd(*remaining)])
                score = sum(i*a for i, a in enumerate(temp, start=1))
                nonlocal max_score
                if score > max_score:
                    max_score = score
            else:
                for i, r in enumerate(remaining[1:], start=1):
                    dfs(remaining[1:i] + remaining[i+1:], temp + [gcd(remaining[0], r)])
            
        dfs(nums, [])
        return max_score